src/Azure/Akka.Streams.Azure.EventHub.V4/EventHubSource.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                _openCallback = GetAsyncCallback<(TaskCompletionSource<NotUsed>, PartitionContext)>(OnOpen);
+                _closeCallback = GetAsyncCallback<(TaskCompletionSource<NotUsed>, PartitionContext, CloseReason)>(OnClose);
+                _processCallback = GetAsyncCallback<ProcessContext>(OnProcessEvents);
+                _errorCallback = GetAsyncCallback<(TaskCompletionSource<NotUsed>, PartitionContext, Exception)>(OnError);
+                _started.CompareAndSet(false, true);
+            }
+
+            public async Task OpenAsync(PartitionContext context)
+            {
+                // It's possible that PreStart wasn't called before this code executes
+                while (!_started)
+                    await Task.Delay(500);
+
+                var completion = new TaskCompletionSource<NotUsed>();
+                _openCallback((completion, context));
+                await completion.Task;
+            }
+
+            private void OnOpen((TaskCompletionSource<NotUsed>, PartitionContext) args)
+            {
+                var (completion, context) = args;
+                
+                if(Log.IsDebugEnabled)
+                    Log.Debug("Partition initializing. Partition: [{0}]", context.ToString());
+                
+                // We need to count the partitions to close the stage only on the last close call,
+                // otherwise further calls to the _closeCallback wouldn't be handled because they
+                // are moved to DeadLetters and the close task is never completed
+                _partitionCount++;
+                completion.TrySetResult(NotUsed.Instance);
+            }
+
+            public async Task CloseAsync(PartitionContext context, CloseReason reason)
+            {
+                var completion = new TaskCompletionSource<NotUsed>();
+                _closeCallback((completion, context, reason));
+                await completion.Task;
+
+                if (_source._createCheckpointOnClose)
+                    await context.CheckpointAsync();
+            }
+
+            private void OnClose((TaskCompletionSource<NotUsed>, PartitionContext, CloseReason) args)
+            {
+                var (completion, context, reason) = args;
+                if(Log.IsDebugEnabled)
+                    Log.Debug("Partition closing. Partition: [{0}], Reason: [{1}]", context.ToString(), reason);
+                
+                if(--_partitionCount == 0)
+                    CompleteStage();
+                completion.TrySetResult(NotUsed.Instance);
+            }
+
+            public async Task ProcessErrorAsync(PartitionContext context, Exception error)
+            {
+                var completion = new TaskCompletionSource<NotUsed>();
+                _errorCallback((completion, context, error));
+                await completion.Task;
+            }
+
+            private void OnError((TaskCompletionSource<NotUsed>, PartitionContext, Exception) args)
```

Comment:
New code, V4 introduces an error callback, need someone to review this and check to see if the error handling behaviour is correct.